Sources/ContainerdRegistry/Registry.swift

```diff
@@ -0,0 +1,1192 @@
+        else {
+            return .notFound(.init(body: .json(.init())))
+        }
+
+        let range = rangeHeaderValue(forUploadedBytes: session.offset)
+
+        return .noContent(.init(headers: .init(range: range)))
+    }
+
+    public func initiateBlobUpload(
+        _ input: OCIRegistryOpenAPI.Operations.InitiateBlobUpload.Input
+    ) async throws -> OCIRegistryOpenAPI.Operations.InitiateBlobUpload.Output {
+        let repositoryName = input.path.name
+        let uuid = UUID().uuidString.lowercased()
+
+        let logger = Logger(label: "sh.wendyengineer.containerd-registry.initiate-blob-upload")
+        let contentClient = Containerd_Services_Content_V1_Content.Client(wrapping: self.client)
+
+        Task {
+            try await contentClient.write { writer in
+                let (stream, continuation) = AsyncStream<Void>.makeStream()
+                await Self.uploadSessions.insert(
+                    uuid: uuid,
+                    repository: repositoryName,
+                    ref: uuid,
+                    write: { message in
+                        do {
+                            try await writer.write(message)
+                        } catch {
+                            logger.error(
+                                "Failed to write chunk",
+                                metadata: ["error": "\(error)", "session-id": .string(uuid)]
+                            )
+                            continuation.finish()
+                            throw error
+                        }
+                    },
+                    finish: {
+                        logger.info(
+                            "Finished writing blob upload session",
+                            metadata: ["session-id": .string(uuid)]
+                        )
+                        continuation.finish()
+                    }
+                )
+
+                for await _ in stream {}
+            } onResponse: { response in
+                do {
+                    for try await _ in response.messages {}
+                    return ()
+                } catch {
+                    logger.error(
+                        "Failed to read response",
+                        metadata: ["error": "\(error)", "session-id": .string(uuid)]
+                    )
+                    throw error
+                }
+            }
+        }
+
+        return .accepted(
+            .init(
+                headers: .init(
+                    location: uuid,
+                    range: rangeHeaderValue(forUploadedBytes: 0)
+                )
+            )
+        )
+    }
+
+    public func getBlobUploadStatus(
+        _ input: OCIRegistryOpenAPI.Operations.GetBlobUploadStatus.Input
+    ) async throws -> OCIRegistryOpenAPI.Operations.GetBlobUploadStatus.Output {
+        let repositoryName = input.path.name
+
+        guard let session = await Self.uploadSessions.firstSession(for: repositoryName) else {
```

Comment:
firstSession here: is it possible that there would be multiple per repository name? and if so, is getting the first one accurate? is uploadSessions sorted in any way?

Reply:
It's not sorted really, so yes that's a bug that could happen. We should have a check for that so we don't have two parallel uploads for the same blob